ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala

```diff
@@ -24,111 +21,166 @@ object MetricHandle {
-    def getCount: Long = metric.getCount
-    def getSnapshot: Snapshot = metric.getSnapshot
-    def getMeanRate: Double = metric.getMeanRate
-    def time(): Context = metric.time()
+  object Timer {
+    type TimerStop = () => Unit
   }
 
-  sealed case class Gauge[U <: codahale.Gauge[T], T](name: String, metric: U)
-      extends MetricHandle[codahale.Gauge[T]] {
+  sealed case class DropwizardTimer(name: String, metric: codahale.Timer) extends Timer {
+
+    def update(duration: Long, unit: TimeUnit): Unit = metric.update(duration, unit)
+    override def time[T](call: => T): T = metric.time(() => call)
+    override def startAsync(): TimerStop = {
+      val ctx = metric.time()
+      () => {
+        ctx.stop()
+        ()
+      }
+    }
+  }
+  sealed trait Gauge[T] extends MetricHandle {
     def metricType: String = "Gauge"
+
+    def updateValue(newValue: T): Unit
+
+    def getValue: T
+  }
+
+  sealed case class DropwizardGauge[T](name: String, metric: Gauges.VarGauge[T]) extends Gauge[T] {
+    def updateValue(newValue: T): Unit = metric.updateValue(newValue)
+    override def getValue: T = metric.getValue
   }
 
-  sealed case class Meter(name: String, metric: codahale.Meter)
-      extends MetricHandle[codahale.Meter] {
+  sealed trait Meter extends MetricHandle {
     def metricType: String = "Meter"
 
-    def mark(): Unit = metric.mark()
+    def mark(): Unit = mark(1)
+    def mark(value: Long): Unit
+
+  }
+  sealed case class DropwizardMeter(name: String, metric: codahale.Meter) extends Meter {
+
+    def mark(value: Long): Unit = metric.mark(value)
 
   }
 
-  sealed case class Counter(name: String, metric: codahale.Counter)
-      extends MetricHandle[codahale.Counter] {
-    def metricType: String = "Counter"
+  sealed trait Counter extends MetricHandle {
+
+    override def metricType: String = "Counter"
+    def inc(): Unit
+    def inc(n: Long): Unit
+    def dec(): Unit
+    def dec(n: Long): Unit
+    def getCount: Long
+  }
+  sealed case class DropwizardCounter(name: String, metric: codahale.Counter) extends Counter {
 
-    def inc(): Unit = metric.inc
-    def inc(n: Long): Unit = metric.inc(n)
-    def dec(): Unit = metric.dec
-    def dec(n: Long): Unit = metric.dec(n)
+    override def inc(): Unit = metric.inc
+    override def inc(n: Long): Unit = metric.inc(n)
+    override def dec(): Unit = metric.dec
+    override def dec(n: Long): Unit = metric.dec(n)
 
-    def getCount: Long = metric.getCount
+    override def getCount: Long = metric.getCount
   }
 
-  sealed case class Histogram(name: String, metric: codahale.Histogram)
-      extends MetricHandle[codahale.Histogram] {
+  sealed trait Histogram extends MetricHandle {
+
     def metricType: String = "Histogram"
+    def update(value: Long): Unit
+    def update(value: Int): Unit
 
-    def update(value: Long): Unit = metric.update(value)
-    def update(value: Int): Unit = metric.update(value)
-    def getSnapshot: Snapshot = metric.getSnapshot
```

Comment:
Needed by canton.

Reply:
We should not depend on the internals of the metrics themselves. If required for testing we can create assertions that work with metrics.